ecommerce/core/utils.py

```diff
@@ -72,3 +79,300 @@ def use_read_replica_if_available(queryset):
+    pattern = r'(key|number|org|start):\((.*?)\)'
+
+    def process_values(values):
+        # Split on AND/OR (case-insensitive) while keeping delimiters
+        tokens = re.split(r'(\s+(?i:AND|OR)\s+)', values)
+
+        # Track changed operators
+        changed_operators = []
+
+        # Process each token
+        processed_tokens = []
+        for token in tokens:
+            token_upper = token.strip().upper()
+            if token_upper == "AND":
+                processed_tokens.append(" OR ")
+                changed_operators.append("AND → OR")
+            elif token_upper == "OR":
+                processed_tokens.append(" AND ")
+                changed_operators.append("OR → AND")
+            else:
+                # Remove '-' from values
+                processed_tokens.append(token.replace("-", ""))
+
+        return "".join(processed_tokens), changed_operators
+
+    modified_string = input_string
+    all_changed_operators = []
+
+    for match in re.finditer(pattern, input_string):
+        category, values = match.groups()
+        if "-" in values:  # Only modify if values contain '-'
+            modified_values, changed_operators = process_values(values)
+            all_changed_operators.extend(changed_operators)
+            modified_string = modified_string.replace(match.group(0), f"{category}:({modified_values})")
+
+    return modified_string, all_changed_operators
+
+
+def extract_course_info(course_string):
+    # Define the regex pattern to extract org, number, and courserun
+    pattern = r'course-v1:([a-zA-Z0-9_]+)\+([a-zA-Z0-9_\.]+)\+([a-zA-Z0-9_]+)'
+
+    # Search for the pattern in the input string
+    match = re.search(pattern, course_string)
+
+    if match:
+        org, number, courserun = match.groups()
+        return {
+            'org': org,
+            'number': number,
+            'courserun': courserun
+        }
+
+    return "Invalid course format"
+
+
+def process_number_value(course_ids):
+    course_run_parsed_info = extract_course_info(course_ids[0])
+    return f'product.key in ({course_run_parsed_info["org"]}+{course_run_parsed_info["number"]})'
+
+
+def fetch_catalog_course_runs(query, limit, site_configuration):
+
+    from ecommerce.coupons.utils import get_catalog_course_runs  # pylint: disable=import-outside-toplevel
+
+    partner = Partner.objects.filter(short_code='edX').get()
+    try:
+        response = get_catalog_course_runs(site=partner.default_site, query=query, limit=limit, offset=0)
+        results = response['results']
+        course_ids = [result['key'] for result in results]
+        return course_ids
+    except (ReqConnectionError, RequestException, Timeout):
+        logger.error('Unable to connect to Catalog API.')
+        return []
+
+
+def create_predicate_from_course_run(course_rns, all_changed_operators):
+    # Remove duplicates and sort for consistency
+    unique_courses = sorted(set(course_rns))
```

Comment:
sorted is an extra operation here
```suggestion
    unique_courses = set(course_rns)
```

Reply:
Done